### ISSUE TITLE
Tests against provided file descriptor fail at random

### ISSUE THREAD

**Issue by silkentrance** [raszi/node-tmp#139]
_testFileGenerated from _fileTest in test/file-test.js and possibly also test/file-sync-test.js will fail at random due to the file descriptor no longer being valid.

This only occurs during automated Travis builds and I was unable to reproduce the issue locally.

The issue seems to be independent from the node version in use and was observed under node v0.8 and also v4.4 and others as well. Restarting the build will normally cause the build to succeed so it cannot be reproduced on Travis as well.

**Comment by silkentrance** [comment]
@raszi 

I have rearranged the code in test/test-file.js#_testFileGenerated a little bit and this seems to have done the trick.

It seems as if the use of fs.statSync will sometimes cause the existing file descriptor to be closed prematurely.

It may also be due to the fact that we use the asynchronous fs operations for creating the file descriptor and then use synchronous operations for testing against it.

This would make all the more sense as we use similar code in test/file-sync-test.js and that has not been rearranged nor did it expose the flaky behavior.

However, I was only able to see failing builds up until node 5.5, so this might have been fixed since then.

**Comment by silkentrance** [comment]
This also affects 

```
      Error: EBADF: bad file descriptor, fstat 
      at Error (native) 
      at Object.fs.fstatSync (fs.js:839:18) 
      at Object._testFileAfterDetachDescriptorImpl (/home/travis/build/raszi/node-tmp/test/file-test.js:58:19) 
```

This time it happened on node v5.8

**Comment by silkentrance** [comment]
@raszi the fd thing is totally random. even after the rearrangement of the code, it will now fail with node v5.10 when fstat'ing the file descriptor.

**Comment by silkentrance** [comment]
Perhaps it is the mixed use of async/sync operations then?

**Comment by silkentrance** [comment]
After having reworked the tests a little bit, it shows that when using multiple vows contexts, e.g.

```
function _testFile(mode, fdTest) {
  return {
    'where fstat is': {
      topic: function (name, fd) {
        console.log('fstat:' + name);
        console.log(fd);
        fs.fstat(fd, this.callback);
      },
      'as expected': function (stat) {
        console.log(stat);
        Test.testStat(stat, mode);
      },
    },
    '2where fstat is': {
      topic: function (name, fd) {
        console.log('fstat:' + name);
        console.log(fd);
        fs.fstat(fd, this.callback);
      },
      'as expected': function (stat) {
        console.log(stat);
        Test.testStat(stat, mode);
      },
    },
  };
}
```

and

```
vows.describe('File creation').addBatch({
  'when using without parameters': {
    topic: function () {
      tmp.file(this.callback);
    },

    'should not return with an error': assert.isNull,
    'should return with a name': Test.assertName,
    'there should be a file': _testFile(0100600, true),
    'should have the default prefix': Test.testPrefix('tmp-'),
    'should have the default postfix': Test.testPostfix('.tmp'),
    'should have been cleaned up': Test.testCleanup
  },
});
```

the file descriptor will no longer be valid in `2where ...`.

It seems as if these file descriptors get closed prematurely.

**Comment by silkentrance** [comment]
The latest build https://travis-ci.org/raszi/node-tmp/builds/251584105 exploits this based on the above changes.

I think that we should move to a different testing framework as proposed in #137.

**Comment by silkentrance** [comment]
I think that I finally nailed this down to the removeCallback being called before that existing vows/tests complete. In the removeCallback, the open file descriptor will be closed and thus it will become unavailable at a random point in time.

Vows, it seems, will not run the tests in order, and sometimes tests that would call upon the removeCallback will be run prior to tests that rely upon a working descriptor.

**Comment by silkentrance** [comment]
This might be fixed when moving to a different testing framework #137.

**Comment by silkentrance** [comment]
fixed with #137
